compiler/rustc_passes/src/check_attr.rs

```diff
@@ -717,6 +717,53 @@ impl CheckAttrVisitor<'tcx> {
         true
     }
 
+    /// Checks that `doc(test(...)` attribute contains only valid attributes. Returns `true` if
+    /// valid.
+    fn check_test_attr(&self, meta: &NestedMetaItem, hir_id: HirId) -> bool {
+        let mut is_valid = true;
+        if let Some(metas) = meta.meta_item_list() {
+            for i_meta in metas {
+                match i_meta.name_or_empty() {
+                    sym::attr | sym::no_crate_inject => {}
+                    _ => {
+                        self.tcx.struct_span_lint_hir(
+                            UNUSED_ATTRIBUTES,
+                            hir_id,
+                            i_meta.span(),
+                            |lint| {
+                                lint.build(&format!(
+                                    "unknown `doc(test)` attribute `{}`",
+                                    rustc_ast_pretty::pprust::path_to_string(
+                                        &i_meta.meta_item().unwrap().path
+                                    ),
+                                ))
+                                .warn(
+                                    "this was previously accepted by the compiler but is \
+                                     being phased out; it will become a hard error in \
+                                     a future release!",
+                                )
+                                .emit();
+                            },
+                        );
+                        is_valid = false;
+                    }
+                }
+            }
+        } else {
+            self.tcx.struct_span_lint_hir(UNUSED_ATTRIBUTES, hir_id, meta.span(), |lint| {
+                lint.build("`#[doc(test(...)]` takes a list of attributes")
+                    .warn(
+                        "this was previously accepted by the compiler but is \
+                         being phased out; it will become a hard error in \
```

Comment:
This is not how future incompat lints work. The diagnostic machinery already prints that text,, this will repeat it. The problem is you used UNUSED_ATTRIBUTES which is not a future incompat lint - please use INVALID_DOC_ATTRS instead. There are examples above and below not 100 lines from here, you don't have to rewrite this from scratch.

Reply:
I'm really completely out of it. Sorry for the mess...